### PR TITLE
Update model docs to reStructuredText format II

### LINKED ISSUE #1354: Update voltmeter documentation
The documentation needs to converted to pure restructured text and updated to the NESTio standards once #1069 is merged to master and master merged to nest-3.

### LINKED ISSUE #1498: documentation generation dependencies
**Describe the bug**
the `doc/conf.py` contains dependencies to `recommonmark.parser` and `copybutton.js`, that are probably not needed anymore and were remarked to be removed, see #1355 

**Expected behavior**
No unnecessary dependencies should be required.

### PR DESCRIPTION
**Pull request by sarakonradi**
As discussed with @jougs, this PR takes over #1374 and includes all the work done by @jessica-mitchell and @terhorstd. 

As described in #1374, the PR: 
- Removes the doxygen style format and replaces it with RST in the ``models`` directory
- Deletes the SLI examples

I have now: 
- Included the updated rate-based model files (#1484)
- Ensured the new MUSIC proxy documentation is properly converted to RST (#1245)
- Fixed any remaining merge conflicts

As noted in the previous PR, we have now lost the automated linking (like in the See Also section) that breathe used to do. Moreover, we should take into account #1355.

fixes #1354 
fixes #1498

### CODE REVIEW

**Comment by sarakonradi** [comment]
@jougs Just merged sarakonradi/nest-simulator#3. Could you please have another look? Thanks!

File: models/iaf_psc_exp.h [hunk 1]
```diff
@@ -64,7 +63,7 @@ equation represents a piecewise constant external current.
 
 The general framework for the consistent formulation of systems with
 neuron like dynamics interacting by point events is described in
-[2]. A flow chart can be found in [3].
+[2]_. A flow chart can be found in [3].
```
Comment (heplesser) [inline models/iaf_psc_exp.h:66]: Should there be a `_` also after `[3]`?

File: models/iaf_psc_exp.h [hunk 2]
```diff
@@ -85,11 +84,25 @@ optimization levels. A future version of iaf_psc_exp will probably
...
+For details, please see IAF_neurons_singularity.ipynb in the
+NEST source code (docs/model_details).
+
+iaf_psc_exp can handle current input in two ways: Current input
+through receptor_type 0 are handled as stepwise constant current
+input as in other iaf models, i.e., this current directly enters
+the membrane potential equation. Current input through
+receptor_type 1, in contrast, is filtered through an exponential
+kernel with the time constant of the excitatory synapse,
+tau_syn_ex. For an example application, see [6].
```
Comment (heplesser) [inline models/iaf_psc_exp.h:99]: You are pointing here to reference [6] (without trailing `_`), but the reference list has only four entries. Maybe some old version of the model reveals what [6] pointed to once upon a time ;).

**Comment by clinssen** [comment]
Do ``pynest/examples/aeif_cond_beta_multisynapse.py`` and ``pynest/examples/aeif_cond_beta_multisynapse.py`` belong in this PR or were they accidently picked up? I cannot see where they come from.

File: models/iaf_psc_exp.h (hunk 2 above)
Comment (sarakonradi) [reply models/iaf_psc_exp.h]: I checked the file's history, and it seems to have been pointing to ``[4]`` in the past: https://github.com/nest/nest-simulator/blob/853911abebdd3d893099f719c4db1d35d0353c96/models/iaf_psc_exp.h

File: models/iaf_psc_exp.h (hunk 1 above)
Comment (sarakonradi) [reply models/iaf_psc_exp.h]: Yes, there should be! Done.

**Comment by sarakonradi** [comment]
@clinssen It seems that the ``pynest/examples`` were added with my merge of @jougs's PR: https://github.com/nest/nest-simulator/pull/1530/commits/a5649518b0fd8287261f0f9437d52a4fab6b45d0

What's the best way to fix this?

**Comment by clinssen** [comment]
The easiest is probably to remove the file and ``git add -u`` and push. To do it "properly" would require removing the file from the PR altogether. See here for a suggested approach: https://www.deployhq.com/git/faqs/removing-large-files-from-git-history

In any case, I think we would all be sufficiently satisfied with the easy approach!

**Comment by jougs** [comment]
@sarakonradi, @clinssen: Please don't remove them!

The examples stem from code snippets in the corresponding C++ header files, which don't belong there. I thus moved them to to being standalone files in the `pynest/examples` directory.

**Comment by clinssen** [comment]
Thanks for the clarification!

**Comment by sarakonradi** [comment]
@jougs Sure, thanks for clarifying the confusion!

**Comment by sarakonradi** [comment]
@heplesser I updated the references, as you suggested. Please check!

**Comment by heplesser** [comment]
> To make this really useful, I think we should wait with the merge until the extension for the "Short description" header by @terhorstd has landed in master and master has been merged here.

@jougs Does that extension have a PR already? I think it would be good to merge this PR soon. Until this PR is fixed, making any new contributions to model documentation is difficult/pending.

**Comment by jougs** [comment]
@heplesser: as far as I know, @terhorstd has the code mostly ready and is now working towards a PR. If we merge as is, we won't get any model documentation for *master*.

**Comment by heplesser** [comment]
> @heplesser: as far as I know, @terhorstd has the code mostly ready and is now working towards a PR. If we merge as is, we won't get any model documentation for _master_.

@jougs OK, we certainly want documentation from master ;).

**Comment by sarakonradi** [comment]
@heplesser @jougs 

> I think it would be good to merge this PR soon. Until this PR is fixed, making any new contributions to model documentation is difficult/pending.

I agree! @terhorstd and I were just discussing this offline and going through the ``conf.py``. I am now doing a final review of sarakonradi/nest-simulator#5. Once that is merged, @terhorstd will send me his PR, and then we should be good to go. :)

**Comment by jougs** [comment]
@sarakonradi: I think you went quite a bit over the top with your changes to the C++ files. As far as I can see, the changes remove essential parts of the functionality and thus render the models unusable. In most cases, formatting changes should only remove/add whitespace and break/unbreak lines.

My suggestion is that you do a `git checkout b2e872d` in your local branch to go back in history, autoformat the files by running `extras/format_all_c_c++_files.sh` and then `git push --force` your changes to GitHub, overwriting the version of the history containing your recent commits 617eaf7, 8811268, 873b04a, and 7d33bba .

**Comment by sarakonradi** [comment]
@jougs You are absolutely right, sorry if I touched too much! I reset the files to their state in https://github.com/nest/nest-simulator/pull/1530/commits/b2e872dee38d4bceeed7be805bb5c5eedf95d619 and will autoformat the C++ files.

**Comment by jougs** [comment]
Something is fishy here. Why did the formatting change 188 files and not just the four which failed the checks previously? I can have a look later today if you like.

**Comment by sarakonradi** [comment]
@jougs Essentially, ``extras/format_all_c_c++_files.sh`` went through all the model files, not just the four files that failed previously. Perhaps the problem is related to my ``clang-format`` version, namely 6.0? If it helps, I can autoformat the files again and manually add only those four files. 

> I can have a look later today if you like.

That would be great!

**Comment by jougs** [comment]
Ah! Your `clang-format` version might already be the source of the problem. We strictly require version 3.6, as different versions are producing incompatible results and we had to settle on one.

Can you please install that and re-check? If it is not available anymore for your OS version, you can install it from scratch using [the same recipe as Travis uses](https://github.com/nest/nest-simulator/blob/master/extras/travis_build.sh#L163).

If it works, please go back in history again and add a new commit with just the new set of changes.

**Comment by sarakonradi** [comment]
> Ah! Your clang-format version might already be the source of the problem. We strictly require version 3.6, as different versions are producing incompatible results and we had to settle on one.

@jougs I see! I now removed version 6.0 and installed 3.6. This time it autoformatted only two of the files, namely ``models/ac_generator.h`` and ``models/gif_pop_psc_exp.h``.  However, it did not touch the other two files. Do you get the same result?

**Comment by jougs** [comment]
This is really stupid... I think I (manually) found and fixed the failing lines. If Travis agrees, I'll send you a PR. If not, I'll investigate further.

**Comment by sarakonradi** [comment]
@jougs Just merged sarakonradi/nest-simulator#7; feel free to have another look once the checks have completed.

**Comment by sarakonradi** [comment]
@jougs Looks like all checks have passed now!

File: models/sinusoidal_gamma_generator.h [hunk 3]
```diff
@@ -140,8 +147,8 @@ SeeAlso: sinusoidal_poisson_generator, gamma_sup_generator
  * are changed at
  *       \f$t_c > t_0\f$, then \f$\Lambda(t)\f$ is integrated piecewise for
  *       \f$t>t_c\f$ as
- *       @f[ \Lambda(t) = a_{old} \int_{t_0}^{t_c]} \lambda_{old}(s) ds
- *                      + a_{new} \int_{t_c}^{t]} \lambda_{new}(s) ds @f]
+ *        \Lambda(t) = a_{old} \int_{t_0}^{t_c]} \lambda_{old}(s) ds
+ *                      + a_{new} \int_{t_c}^{t]} \lambda_{new}(s) ds
```
Comment (jougs) [inline models/sinusoidal_gamma_generator.h:151]: These are all inside of a Doxygen comment and should thus not be changed to RST syntax. Please revert.

File: models/spike_detector.h [hunk 4]
```diff
@@ -33,11 +33,14 @@
 #include "nest_types.h"
 #include "recording_device.h"
 
-/* BeginUserDocs: device, recorder, spikes
+/* BeginUserDocs: device, recorder, spike
 
 Collecting spikes from neurons
 ##############################
```
Comment (jougs) [inline models/spike_detector.h:39]:
```suggestion
Short description
+++++++++++++++++

Collecting spikes from neurons
```

File: models/spike_dilutor.h [hunk 5]
```diff
@@ -34,13 +34,20 @@
...
-@ingroup Devices
-@ingroup generator
+/* BeginUserDocs: device, generator
 
-Name: spike_dilutor - repeats incoming spikes with a certain probability.
+Short description
++++++++++++++++++
 
-Description:
+repeats incoming spikes with a certain probability
```
Comment (jougs) [inline models/spike_dilutor.h:42]:
```suggestion
Repeat incoming spikes with a certain probability
```

File: models/spin_detector.h [hunk 6]
```diff
@@ -75,10 +77,17 @@ be recorded, are recorded.
...
 states are not necessarily written to file in chronological order.
 
-Receives: SpikeEvent
+Receives
+++++++++
+
+SpikeEvent
+
+See also
+++++++++
```
Comment (jougs) [inline models/spin_detector.h:86]:
```suggestion
```

File: models/spin_detector.h [hunk 7]
```diff
@@ -75,10 +77,17 @@ be recorded, are recorded.
...
 
-SeeAlso: spike_detector, Device, RecordingDevice
-*/
+spike_detector, Device, RecordingDevice
+
```
Comment (jougs) [inline models/spin_detector.h:89]:
```suggestion
```

File: models/stdp_connection.h [hunk 8]
```diff
@@ -71,13 +66,16 @@ exponent can be set separately for potentiation and depression.
...
-Transmits: SpikeEvent
+Transmits
++++++++++
+
+SpikeEvent
+
+References
+++++++++++
 
-References:
```
Comment (jougs) [inline models/stdp_connection.h:78]:
```suggestion
```

File: models/stdp_connection.h [hunk 9]
```diff
@@ -92,18 +90,17 @@ Transmits: SpikeEvent
...
        20(23):8812-8821.
        DOI: https://doi.org/10.1523/JNEUROSCI.20-23-08812.2000
-\endverbatim
 
-FirstVersion: March 2006
+See also
+++++++++
 
-Author: Moritz Helias, Abigail Morrison
+synapsedict, tsodyks_synapse, static_synapse
```
Comment (jougs) [inline models/stdp_connection.h:97]:
```suggestion
tsodyks_synapse, static_synapse
```

File: models/stdp_triplet_connection.h [hunk 10]
```diff
@@ -30,81 +30,81 @@
...
-  can be set as properties of the postsynaptic neuron.
-- This version implements the 'all-to-all' spike interaction of [1]. The
-  'nearest-spike' interaction of [1] can currently not be implemented
-  without changing the postsynaptic archiving-node (clip the traces to a
-  maximum of 1).
+See also
+++++++++
 
-FirstVersion: Nov 2007
+stdp_triplet_synapse_hpc, synapsedict, stdp_synapse, static_synapse
```
Comment (jougs) [inline models/stdp_triplet_connection.h:100]:
```suggestion
stdp_triplet_synapse_hpc, stdp_synapse, static_synapse
```

File: models/step_current_generator.h [hunk 11]
```diff
@@ -39,60 +39,54 @@
...
-    sc << /amplitude_times [0.2 0.5] /amplitude_values [2.0 4.0] >> SetStatus
+CurrentEvent
 
-    The amplitude of the DC will be 0.0 pA in the time interval [0, 0.2),
-    2.0 pA in the interval [0.2, 0.5) and 4.0 from then on.
+See also
+++++++++
 
-Sends: CurrentEvent
+ac_generator, dc_generator, step_current_generator, Device, StimulatingDevice
```
Comment (jougs) [inline models/step_current_generator.h:86]:
```suggestion
ac_generator, dc_generator, step_current_generator
```

File: models/weight_recorder.h [hunk 12]
```diff
@@ -36,8 +36,11 @@
 
 /* BeginUserDocs: device, recorder
 
-Recording weights from synapses
-###############################
+weight_recorder - Recording weights from synapses
+#################################################
```
Comment (jougs) [inline models/weight_recorder.h:40]:
```suggestion
Short description
+++++++++++++++++

Recording weights from synapses
```

**Comment by jougs** [comment]
> @jougs Looks like all checks have passed now!

Just for the record: I get automatically notified upon all changes within issues/PRs I was mentioned in. So there's no need to tell me manually. That said: thanks for the notification :-)

File: models/correlation_detector.h [hunk 13]
```diff
@@ -146,33 +146,18 @@ See Auto- and crosscorrelation functions for spike
...
-Receives: SpikeEvent
+SpikeEvent
 
-Author: Moritz Helias
-        Jakob Jordan (implemented Kahan summation algorithm) 2013/02/18
+See also
+++++++++
 
-FirstVersion: 2007/5/21
+spike_detector, Device, PseudoRecordingDevice
```
[the remark this answers is not in the document]
Comment (sarakonradi) [reply models/correlation_detector.h:157]: Sure, I have made a note.

File: models/correlomatrix_detector.h [hunk 14]
```diff
@@ -77,9 +84,29 @@ The correlomatrix_detector has a variable number of inputs which can be set
...
 
-Parameters:
+Remarks:
+
+This recorder does not record to file, screen or memory in the usual
+sense.
+
+@note Correlomatrix detectors IGNORE any connection delays.
+
+@note Correlomatrix detector breaks with the persistence scheme as
```
[the remark this answers is not in the document]
Comment (sarakonradi) [reply models/correlomatrix_detector.h:94]: I will have a look.

**Comment by sarakonradi** [comment]
Thanks for the commit suggestions! Just applied them. :) 

I built the documentation locally, and the math renders well (see below). However, it is good practice to indent it regardless. 

Screenshot from 2020-05-13 16-28-34

I will have a further look at the ``.. math::``, "Remarks" and Doxygen comments.

**Comment by sarakonradi** [comment]
Merged sarakonradi/nest-simulator#6 after @jougs approved offline.

**Comment by jougs** [comment]
> Thanks for the commit suggestions! Just applied them. :)

There's still a bunch left, which got hidden by GitHub.

> I will have a further look at the `.. math::`, "Remarks" and Doxygen comments.

If it renders well, I'd leave this for another time. This PR really needs to be merged to master, so we can build new things on top of it. Let's just pretend we did not see this ;-)

**Comment by sarakonradi** [comment]
> There's still a bunch left, which got hidden by GitHub.

Ah, it's never-ending. Just committed those too.

> If it renders well, I'd leave this for another time. This PR really needs to be merged to master, so we can build new things on top of it. Let's just pretend we did not see this ;-)

I have taken a note of your comments. If you prefer, we can merge this PR first and work on those enhancements in a follow-up PR. :)

**Comment by jougs** [comment]
> Ah, it's never-ending. Just committed those too.

And now tests are failing again. I think we really deserve a drink after this one...

Can you please address the issues listed [in the Travis log](https://travis-ci.org/github/nest/nest-simulator/jobs/686627984#L1328)?

# :crying_cat_face: ➟  :cat:, please!

**Comment by sarakonradi** [comment]
Done! 😄 Staticcheck just passed. Hopefully Travis will be happy with the rest too...

**Comment by jougs** [comment]
I'll approve and merge as soon as it is. And then we're not talking about this one anymore. Ever.

**Comment by sarakonradi** [comment]
Yes. Please.

**Comment by jougs** [comment]
Woohoo!

# :cat: :tada: :balloon: :man_dancing:  :woman_dancing:

**Comment by sarakonradi** [comment]
Finally! 🎉

**Pull request merged**